File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int time = 0;
        int fresh = 0;
        Queue<int[]> rotten = new LinkedList<>();
        int m = grid.length;
        int n = grid[0].length;
        for(int i = 0;i < m; i++){
            for(int j = 0;j < n; j++){
                if(grid[i][j] == 1){
                    fresh++;
                }
                if(grid[i][j] == 2){
                    rotten.offer(new int[]{i, j});
                }
            }
        }
        int[][] dir = {
            {0,1},
            {0,-1},
            {1,0},
            {-1,0}
        };
        while(!rotten.isEmpty() && fresh > 0){
            int size = rotten.size();
            for(int i = 0;i < size; i++){
                int[] cur = rotten.poll();
                for(int[] d : dir){
                    int x = cur[0] + d[0];
                    int y = cur[1] + d[1];
                    if(x >= 0 && y >= 0 && x < m && y < n && grid[x][y] == 1){
                        grid[x][y] = 2;
                        fresh--;
                        rotten.offer(new int[]{x, y});
                    }
                }
            }
            time++;
        }
        return fresh == 0 ? time : -1;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
from collections import deque

class Solution(object):
    def orangesRotting(self, grid):
        fresh, time = 0, 0
        rotten = deque()
        Y = len(grid)
        X = len(grid[0])
        for i in range(Y):
            for j in range(X):
                if grid[i][j] == 1:
                    fresh += 1
                elif grid[i][j] == 2:
                    rotten.append((i, j))
        directions = [(0,1), (0,-1), (1,0), (-1,0)]
        while rotten and fresh > 0:
            for _ in range(len(rotten)):
                y, x = rotten.popleft()
                for dy, dx in directions:
                    ny, nx = y + dy, x + dx
                    if 0 <= ny < Y and 0 <= nx < X and grid[ny][nx] == 1:
                        grid[ny][nx] = 2
                        fresh -= 1
                        rotten.append((ny, nx))
            time += 1
        return time if fresh == 0 else -1